Derivative blocks

glubtier opened the pull request:
BlockStateProviders created for derivative blocks: Walls, Buttons, PressurePlates, Fences, FenceGates, Doors, and Trapdoors
Added derivatives for Minestuck-specific stones and woods
Added Aspect tree blocks to CreativeTabs
Added derivatives to stonecutter recipes where applicable
Added Perfectly Generic blocks
Fixed loot table for double slabs to drop 2 slabs.

All new textures are by glubtier.

kirderf1 commented:
Alright, nice!
There are a couple things of note:
- This seems to be built on top of the branch from #544, which is a bit unfortunate since this PR now has some already-squashed commits. Nothing to do about it now, but something to think about for next time you start on something new.
- You've added some commented-out code in places. I think those should either be removed from this PR or fixed/added this PR. (And if you need help fixing something, feel free to ask!)
- Spacing between function arguments is a little odd in some places. This can easily be fixed with the "Reformat Code" action in changed files with your IDE.
- As seen with the failed check, you should run "runData" to generate the new json files.
- Changelog entry should be added!
- If you could add texture credit in the PR description or a separate comment, that'd be great.

glubtier commented:
- Noted!
- Uncommented and added textures (if anything is still commented out it's because I missed it but I think I got them all.)
- Sure, I went ahead and ran that on the files I updated.
- Noted, didn't know I should commit the generated files but I've done so now!
- Updated changelog
- All new textures are by me/glubtier

kirderf1 commented on src/main/java/com/mraof/minestuck/data/recipe/MinestuckRecipeProvider.java:
```diff
@@ -312,7 +330,6 @@ protected void buildRecipes(Consumer<FinishedRecipe> recipeBuilder)
 		SingleItemRecipeBuilder.stonecutting(Ingredient.of(MSBlocks.COARSE_STONE.get()), RecipeCategory.BUILDING_BLOCKS, MSBlocks.COARSE_STONE_STAIRS.get()).unlockedBy("has_coarse_stone", has(MSBlocks.COARSE_STONE.get())).save(recipeBuilder, new ResourceLocation(Minestuck.MOD_ID, "coarse_stone_stairs_from_coarse_stone_stonecutting"));
 		SingleItemRecipeBuilder.stonecutting(Ingredient.of(MSBlocks.COARSE_STONE.get()), RecipeCategory.BUILDING_BLOCKS, MSBlocks.COARSE_STONE_SLAB.get(), 2).unlockedBy("has_coarse_stone", has(MSBlocks.COARSE_STONE.get())).save(recipeBuilder, new ResourceLocation(Minestuck.MOD_ID, "coarse_stone_slab_from_coarse_stone_stonecutting"));
 		SingleItemRecipeBuilder.stonecutting(Ingredient.of(MSBlocks.COARSE_STONE_BRICKS.get()), RecipeCategory.BUILDING_BLOCKS, MSBlocks.COARSE_STONE_BRICK_STAIRS.get()).unlockedBy("has_coarse_stone_bricks", has(MSBlocks.COARSE_STONE_BRICKS.get())).save(recipeBuilder, new ResourceLocation(Minestuck.MOD_ID, "coarse_stone_brick_stairs_from_coarse_stone_bricks_stonecutting"));
-		SingleItemRecipeBuilder.stonecutting(Ingredient.of(MSBlocks.COARSE_STONE_BRICKS.get()), RecipeCategory.BUILDING_BLOCKS, MSBlocks.COARSE_STONE_BRICK_SLAB.get(), 2).unlockedBy("has_coarse_stone_bricks", has(MSBlocks.COARSE_STONE_BRICKS.get())).save(recipeBuilder, new ResourceLocation(Minestuck.MOD_ID, "coarse_stone_brick_slab_from_coarse_stone_bricks_stonecutting"));
```
Looks like this recipe was removed without a replacement. Was that intentional? Same thing for the frost tile slab and frost brick slab.

glubtier replied:
Not intentional. iirc, I accidentally duplicated those lines, must have gotten lost in trying to remove the dups. Added all three slabs back.

kirderf1 commented on src/main/java/com/mraof/minestuck/data/loot_table/MSBlockLootTables.java:
Most or all lambda expressions here can be simplified into method references. Here's what that means.
```java
this.add(BLACK_CHESS_BRICK_SLAB.get(), block -> createSlabItemTable(BLACK_CHESS_BRICK_SLAB.get()));
```
The block parameter inside the lambda that you give `this.add()` will have the same block as what you gave directly to `this.add()`. Here for example, `block` will conain the black chess brick slab block, since that's the block that goes into `this.add()`. The above expression is then the same as:
```java
this.add(BLACK_CHESS_BRICK_SLAB.get(), block -> createSlabItemTable(block));
```
At this point, your IDE likely notices what's going on and will suggest to use a method reference instead of a lambda. Here's what that would look like:
```java
this.add(BLACK_CHESS_BRICK_SLAB.get(), this::createSlabItemTable);
```

glubtier replied:
Thanks, this was helpful! I updated loottable entries for slabs and doors.

kirderf1 commented on src/main/java/com/mraof/minestuck/data/MSBlockStateProvider.java:
```diff
@@ -261,15 +387,32 @@ protected void registerStatesAndModels()
+		slabWithItem(MSBlocks.FLOWERY_MOSSY_COBBLESTONE_SLAB, MSBlocks.FLOWERY_MOSSY_COBBLESTONE);
+		wallWithItem(MSBlocks.FLOWERY_MOSSY_COBBLESTONE_WALL, MSBlocks.FLOWERY_MOSSY_COBBLESTONE);
+		
+		//TODO update derivatives to use dynamic textures
 		variantsWithItem(MSBlocks.FLOWERY_MOSSY_STONE_BRICKS, 4,
 				i -> models().cubeAll("flowery_mossy_stone_bricks" + (i + 1),
 						texture(id("flowery_mossy_stone_bricks/" + (i + 1)))));
+		//stairsWithItem(MSBlocks.FLOWERY_MOSSY_STONE_BRICK_STAIRS, MSBlocks.FLOWERY_MOSSY_STONE_BRICKS);
+		//slabWithItem(MSBlocks.FLOWERY_MOSSY_STONE_BRICK_SLAB, MSBlocks.FLOWERY_MOSSY_STONE_BRICKS);
+		//wallWithItem(MSBlocks.FLOWERY_MOSSY_STONE_BRICK_WALL, MSBlocks.FLOWERY_MOSSY_STONE_BRICKS);
+		
 		variantsWithItem(MSBlocks.DECREPIT_STONE_BRICKS, 4,
 				i -> models().cubeAll("decrepit_stone_bricks" + (i + 1),
 						texture(id("decrepit_stone_bricks/" + (i + 1)))));
+		//stairsWithItem(MSBlocks.DECREPIT_STONE_BRICK_STAIRS, MSBlocks.DECREPIT_STONE_BRICKS);
+		//slabWithItem(MSBlocks.DECREPIT_STONE_BRICK_SLAB, MSBlocks.DECREPIT_STONE_BRICKS);
+		//wallWithItem(MSBlocks.DECREPIT_STONE_BRICK_WALL, MSBlocks.DECREPIT_STONE_BRICKS);
+		
 		variantsWithItem(MSBlocks.MOSSY_DECREPIT_STONE_BRICKS, 4,
 				i -> models().cubeAll("mossy_decrepit_stone_bricks" + (i + 1),
 						texture(id("mossy_decrepit_stone_bricks/" + (i + 1)))));
+		//stairsWithItem(MSBlocks.MOSSY_DECREPIT_STONE_BRICK_STAIRS, MSBlocks.MOSSY_DECREPIT_STONE_BRICKS);
+		//slabWithItem(MSBlocks.MOSSY_DECREPIT_STONE_BRICK_SLAB, MSBlocks.MOSSY_DECREPIT_STONE_BRICKS);
+		//wallWithItem(MSBlocks.MOSSY_DECREPIT_STONE_BRICK_WALL, MSBlocks.MOSSY_DECREPIT_STONE_BRICKS);
```
This todo and commented code need resolving. The flowery mossy stone brick stairs/slab did have datagen before this PR, which might work as guidance.

glubtier replied:
I intended to remove these and missed them. I can either
1. Remove them like I meant to, and work on them another time
2. Keep working at them now, at the expense of getting all the other blocks available for builders

Whatever you think is best!

kirderf1 replied:
I'd lean towards getting them done now, but someone else can be the judge for how quickly this need to be merged and ready for builders.
But at the very least, the flowery mossy stone brick stairs/slab **should** be handled now since they were already present before this PR.

glubtier replied:
Okay so when I look at FLOWERY_MOSSY_STONE_BRICK_STAIRS and FLOWERY_MOSSY_STONE_BRICK_SLAB I see that they're just using one texture (the first texture), i.e.:

`stairsWithItem(MSBlocks.FLOWERY_MOSSY_STONE_BRICK_STAIRS, "flowery_mossy_stone_brick", texture("flowery_mossy_stone_bricks/1"));`

So I updated slabs and walls for DECREPIT_STONE_BRICKS and MOSSY_DECREPIT_STONE_BRICKS to be consistent with that. I also modified the slab method to "match" the other two a little more, but I can revert this and go with the old way -- I just thought this looked a little more tidy.

kirderf1 commented on src/main/java/com/mraof/minestuck/block/MSBlocks.java:
```diff
@@ -210,51 +299,90 @@ private static Block coalOre(BlockBehaviour.Properties properties)
+	public static final RegistryObject<WallBlock> FLOWERY_MOSSY_COBBLESTONE_WALL = REGISTER.register("flowery_mossy_cobblestone_wall", () -> new WallBlock(copy(FLOWERY_MOSSY_COBBLESTONE.get())));
 	
 	public static final RegistryObject<Block> FLOWERY_MOSSY_STONE_BRICKS = REGISTER.register("flowery_mossy_stone_bricks", () -> new Block(Block.Properties.of().mapColor(MapColor.COLOR_GRAY).instrument(NoteBlockInstrument.BASEDRUM).requiresCorrectToolForDrops().strength(1.5F, 6.0F)));
-	public static final RegistryObject<StairBlock> FLOWERY_MOSSY_STONE_BRICK_STAIRS = REGISTER.register("flowery_mossy_stone_brick_stairs", () -> new StairBlock(() -> MSBlocks.FLOWERY_MOSSY_STONE_BRICKS.get().defaultBlockState(), copy(FLOWERY_MOSSY_STONE_BRICKS.get())));
```
This removal seems a bit odd. With the non-supplier constructor being deprecated, I think it's better to keep to the supplier-based constructor instead.

glubtier replied:
This will depend on whether I'm adding or removing these, I'll revert this if I'm actually adding them.

kirderf1 replied:
Not sure I follow. Is there any undecided question that you're referring to?

glubtier replied:
The previously commented-out lines, but disregard. It's been added back.

kirderf1 replied:
Alright, thanks!
I'd say that the same thing should apply to the new stairs blocks as well. You should be able to spot the deprecations directly in the file in your IDE, or else you could do a search on the deprecated constructor in your IDE.

kirderf1 commented on CHANGELOG.md:
There's also a loot table fix for slabs.

glubtier replied:
I put that under "Fixed": "- Double slabs will now drop two slabs when broken", I can rephrase it though.
